// @medusajs/core-flows/dist/handlers/cart/detach-cart-from-sales-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.detachCartFromSalesChannel = void 0;
const utils_1 = require("@medusajs/utils");
var Aliases;
(function (Aliases) {
    Aliases["Cart"] = "cart";
    Aliases["SalesChannel"] = "sales_channel";
})(Aliases || (Aliases = {}));
async function detachCartFromSalesChannel({ container, data, }) {
    const featureFlagRouter = container.resolve("featureFlagRouter");
    const remoteLink = container.resolve("remoteLink");
    if (!featureFlagRouter.isFeatureEnabled(utils_1.MedusaV2Flag.key)) {
        return;
    }
    const cart = data[Aliases.Cart];
    const salesChannel = data[Aliases.SalesChannel];
    await remoteLink.dismiss({
        cartService: {
            cart_id: cart.id,
        },
        salesChannelService: {
            sales_channel_id: salesChannel.sales_channel_id,
        },
    });
}
exports.detachCartFromSalesChannel = detachCartFromSalesChannel;
detachCartFromSalesChannel.aliases = Aliases;
