// @medusajs/modules-sdk/dist/definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MODULE_DEFINITIONS = exports.ModulesDefinition = exports.MODULE_PACKAGE_NAMES = exports.ModuleRegistrationName = exports.Modules = void 0;
const types_1 = require("@medusajs/types");
const utils_1 = require("@medusajs/utils");
var Modules;
(function (Modules) {
    Modules["LINK"] = "linkModules";
    Modules["EVENT_BUS"] = "eventBus";
    Modules["STOCK_LOCATION"] = "stockLocationService";
    Modules["INVENTORY"] = "inventoryService";
    Modules["CACHE"] = "cacheService";
    Modules["PRODUCT"] = "productService";
    Modules["PRICING"] = "pricingService";
    Modules["PROMOTION"] = "promotion";
    Modules["AUTHENTICATION"] = "authentication";
    Modules["WORKFLOW_ENGINE"] = "workflows";
    Modules["CART"] = "cart";
    Modules["CUSTOMER"] = "customer";
    Modules["PAYMENT"] = "payment";
})(Modules || (exports.Modules = Modules = {}));
var ModuleRegistrationName;
(function (ModuleRegistrationName) {
    ModuleRegistrationName["EVENT_BUS"] = "eventBusModuleService";
    ModuleRegistrationName["STOCK_LOCATION"] = "stockLocationService";
    ModuleRegistrationName["INVENTORY"] = "inventoryService";
    ModuleRegistrationName["CACHE"] = "cacheService";
    ModuleRegistrationName["PRODUCT"] = "productModuleService";
    ModuleRegistrationName["PRICING"] = "pricingModuleService";
    ModuleRegistrationName["PROMOTION"] = "promotionModuleService";
    ModuleRegistrationName["AUTHENTICATION"] = "authenticationModuleService";
    ModuleRegistrationName["WORKFLOW_ENGINE"] = "workflowsModuleService";
    ModuleRegistrationName["CART"] = "cartModuleService";
    ModuleRegistrationName["CUSTOMER"] = "customerModuleService";
    ModuleRegistrationName["PAYMENT"] = "paymentModuleService";
})(ModuleRegistrationName || (exports.ModuleRegistrationName = ModuleRegistrationName = {}));
exports.MODULE_PACKAGE_NAMES = {
    [Modules.LINK]: "@medusajs/link-modules",
    [Modules.PRODUCT]: "@medusajs/product",
    [Modules.EVENT_BUS]: "@medusajs/event-bus-local",
    [Modules.STOCK_LOCATION]: "@medusajs/stock-location",
    [Modules.INVENTORY]: "@medusajs/inventory",
    [Modules.CACHE]: "@medusajs/cache-inmemory",
    [Modules.PRICING]: "@medusajs/pricing",
    [Modules.PROMOTION]: "@medusajs/promotion",
    [Modules.AUTHENTICATION]: "@medusajs/authentication",
    [Modules.WORKFLOW_ENGINE]: "@medusajs/workflow-engine-inmemory",
    [Modules.CART]: "@medusajs/cart",
    [Modules.CUSTOMER]: "@medusajs/customer",
    [Modules.PAYMENT]: "@medusajs/payment",
};
exports.ModulesDefinition = {
    [Modules.EVENT_BUS]: {
        key: Modules.EVENT_BUS,
        isLegacy: true,
        registrationName: ModuleRegistrationName.EVENT_BUS,
        defaultPackage: exports.MODULE_PACKAGE_NAMES[Modules.EVENT_BUS],
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.EVENT_BUS),
        canOverride: true,
        isRequired: true,
        dependencies: ["logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.STOCK_LOCATION]: {
        key: Modules.STOCK_LOCATION,
        isLegacy: true,
        registrationName: ModuleRegistrationName.STOCK_LOCATION,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.STOCK_LOCATION),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["eventBusService"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.INVENTORY]: {
        key: Modules.INVENTORY,
        isLegacy: true,
        registrationName: ModuleRegistrationName.INVENTORY,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.INVENTORY),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["eventBusService"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.CACHE]: {
        key: Modules.CACHE,
        isLegacy: true,
        registrationName: ModuleRegistrationName.CACHE,
        defaultPackage: exports.MODULE_PACKAGE_NAMES[Modules.CACHE],
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.CACHE),
        isRequired: true,
        canOverride: true,
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.PRODUCT]: {
        key: Modules.PRODUCT,
        registrationName: ModuleRegistrationName.PRODUCT,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.PRODUCT),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: [ModuleRegistrationName.EVENT_BUS, "logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.PRICING]: {
        key: Modules.PRICING,
        registrationName: ModuleRegistrationName.PRICING,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.PRICING),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.PROMOTION]: {
        key: Modules.PROMOTION,
        registrationName: ModuleRegistrationName.PROMOTION,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.PROMOTION),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.AUTHENTICATION]: {
        key: Modules.AUTHENTICATION,
        registrationName: ModuleRegistrationName.AUTHENTICATION,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.AUTHENTICATION),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.WORKFLOW_ENGINE]: {
        key: Modules.WORKFLOW_ENGINE,
        registrationName: ModuleRegistrationName.WORKFLOW_ENGINE,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.WORKFLOW_ENGINE),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.CART]: {
        key: Modules.CART,
        registrationName: ModuleRegistrationName.CART,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.CART),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.CUSTOMER]: {
        key: Modules.CUSTOMER,
        registrationName: ModuleRegistrationName.CUSTOMER,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.CUSTOMER),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
    [Modules.PAYMENT]: {
        key: Modules.PAYMENT,
        registrationName: ModuleRegistrationName.PAYMENT,
        defaultPackage: false,
        label: (0, utils_1.upperCaseFirst)(ModuleRegistrationName.PAYMENT),
        isRequired: false,
        canOverride: true,
        isQueryable: true,
        dependencies: ["logger"],
        defaultModuleDeclaration: {
            scope: types_1.MODULE_SCOPE.INTERNAL,
            resources: types_1.MODULE_RESOURCE_TYPE.SHARED,
        },
    },
};
exports.MODULE_DEFINITIONS = Object.values(exports.ModulesDefinition);
exports.default = exports.MODULE_DEFINITIONS;
